src/openrct2-ui/windows/TrackDesignPlace.cpp

```diff
@@ -70,633 +70,553 @@ static rct_widget window_track_place_widgets[] = {
         }
     }
-}
 
-void TrackPlaceRestoreProvisional()
-{
-    if (_window_track_place_last_was_valid)
+private:
+    void WindowTrackPlaceClearProvisional()
     {
-        auto tdAction = TrackDesignAction({ _windowTrackPlaceLastValid, _currentTrackPieceDirection }, *_trackDesign);
-        tdAction.SetFlags(GAME_COMMAND_FLAG_NO_SPEND | GAME_COMMAND_FLAG_GHOST);
-        auto res = GameActions::Execute(&tdAction);
-        if (res.Error != GameActions::Status::Ok)
+        if (_window_track_place_last_was_valid)
         {
-            _window_track_place_last_was_valid = false;
+            auto newRide = get_ride(_window_track_place_ride_index);
+            if (newRide != nullptr)
+            {
+                TrackDesignPreviewRemoveGhosts(_trackDesign.get(), newRide, _windowTrackPlaceLastValid);
+                _window_track_place_last_was_valid = false;
+            }
         }
     }
-}
-
-/**
- *
- *  rct2: 0x006D17C6
- */
-static int32_t WindowTrackPlaceGetBaseZ(const CoordsXY& loc)
-{
-    auto surfaceElement = map_get_surface_element_at(loc);
-    if (surfaceElement == nullptr)
-        return 0;
-
-    auto z = surfaceElement->GetBaseZ();
 
-    // Increase Z above slope
-    if (surfaceElement->GetSlope() & TILE_ELEMENT_SLOPE_ALL_CORNERS_UP)
+    int32_t WindowTrackPlaceGetBaseZ(const CoordsXY& loc)
     {
-        z += 16;
+        auto surfaceElement = map_get_surface_element_at(loc);
+        if (surfaceElement == nullptr)
+            return 0;
 
-        // Increase Z above double slope
-        if (surfaceElement->GetSlope() & TILE_ELEMENT_SLOPE_DOUBLE_HEIGHT)
+        auto z = surfaceElement->GetBaseZ();
+
+        // Increase Z above slope
+        if (surfaceElement->GetSlope() & TILE_ELEMENT_SLOPE_ALL_CORNERS_UP)
+        {
             z += 16;
-    }
 
-    // Increase Z above water
-    if (surfaceElement->GetWaterHeight() > 0)
-        z = std::max(z, surfaceElement->GetWaterHeight());
+            // Increase Z above double slope
+            if (surfaceElement->GetSlope() & TILE_ELEMENT_SLOPE_DOUBLE_HEIGHT)
+                z += 16;
+        }
 
-    return z + TrackDesignGetZPlacement(_trackDesign.get(), GetOrAllocateRide(PreviewRideId), { loc, z });
-}
+        // Increase Z above water
+        if (surfaceElement->GetWaterHeight() > 0)
+            z = std::max(z, surfaceElement->GetWaterHeight());
 
-/**
- *
- *  rct2: 0x006CFD9D
- */
-static void WindowTrackPlacePaint(rct_window* w, rct_drawpixelinfo* dpi)
-{
-    auto ft = Formatter::Common();
-    ft.Add<char*>(_trackDesign->name.c_str());
-    WindowDrawWidgets(w, dpi);
+        return z + TrackDesignGetZPlacement(_trackDesign.get(), GetOrAllocateRide(PreviewRideId), { loc, z });
+    }
 
-    // Draw mini tile preview
-    rct_drawpixelinfo clippedDpi;
-    if (clip_drawpixelinfo(&clippedDpi, dpi, w->windowPos + ScreenCoordsXY{ 4, 18 }, 168, 78))
+    void WindowTrackPlaceDrawMiniPreview(TrackDesign* td6)
```

Comment:
For all of the methods that are now part of the class you should drop the `WindowTrackPlace`